### ISSUE TITLE
UnhandledPromiseRejectionWarning

### ISSUE THREAD

**Issue by aviralgoyal** [knex/knex#4128]
# Environment

Knex version: 0.21.5
Database + version: MSSQL 2017
OS: MAC OS

Select applicable template from below.
For MSSql tag @smorey2.
Rest of dialects don't need tags.

# Problem

I am using transactions running in a cron job every 15 seconds hitting the database and performing read, write and update calls bound in **async await try catch block**.

Sample Code:
```
async function updateSetup(tableId) {
    try {
        let newId;
        await dbData.transaction(async trx => {
            let id = await trx('SETUP').select('VALUEZ').where('KEY_NAME', tableId);
            newId = id[0].VALUEZ + 1;
            await trx('SETUP').update({ 'VALUEZ': newId }).where('KEY_NAME', tableId);
        })
        return newId;
    } catch (e) {
        console.log(e);
        throw e;
    }
}

```

**Test**
I was having socket hangup issues and was trying to test the code by switching my internet on and off, to check what all errors will be generated. In one particular case I get this warning  below  and it does not get caught in the catch block:

```
node:41181) UnhandledPromiseRejectionWarning: TypeError: Cannot assign to read only property 'originalError' of object 'TransactionError: Requests can only be made in the LoggedIn state, not the Final state'
    at /Users/aviral/projects/texma-sync copy/node_modules/knex/lib/dialects/mssql/transaction.js:41:38
<node_internals>/internal/process/warning.js:27
(node:41181) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 10)
<node_internals>/internal/process/warning.js:27
(node:41181) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
<node_internals>/internal/process/warning.js:27
``` 

**Issue**
After that  even if I restore the net connection my code resumes after several seconds as this below and pretty much stuck at this point, I have to restart my application manually from PM2 to get rid of this issue:

```
RequestError: select * from [MSGTABLE] where [STATUSZ] = @p0 - Connection lost - socket hang up
    at handleError (/Users/aviral/projects/texma-sync/node_modules/mssql/lib/tedious/request.js:366:15)
    at Connection.emit (<node_internals>/events.js:214:15)
    at Connection.socketError (/Users/aviral/projects/texma-sync/node_modules/tedious/lib/connection.js:1294:12)
    at Connection.socketEnd (/Users/aviral/projects/texma-sync/node_modules/tedious/lib/connection.js:1312:12)
    at Socket.<anonymous> (/Users/aviral/projects/texma-sync/node_modules/tedious/lib/connection.js:1131:14)
    at Socket.emit (<node_internals>/events.js:214:15)
    at endReadableNT (<node_internals>/_stream_readable.js:1178:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21) {code: 'EREQUEST', originalError: Error: Connection lost - socket hang up
  … (/Users/aviral/projects/texma-sync/node_…, name: 'RequestError', number: undefined, lineNumber: undefined, …}
```

I don't know how to resolve this issue, because net connectivity failure is a constant issue at production site.

**Comment by elhigu** [comment]
This sounds like a pretty critical bug.

Could you check pool stats, when everything hangs? 

From `knex.client.pool` 

```
// returns the number of non-free resources
pool.numUsed();

// returns the number of free resources
pool.numFree();

// how many acquires are waiting for a resource to be released
pool.numPendingAcquires();

// how many asynchronous create calls are running
pool.numPendingCreates();
```

Could be that exception in mssql dialect causes transaction to fail without releasing the connection back to the pool, which eventually blocks knex from acquiring any new connections.  

You could try to comment out this line https://github.com/knex/knex/blob/master/lib/dialects/mssql/transaction.js#L41 from your `/Users/aviral/projects/texma-sync copy/node_modules/knex/lib/dialects/mssql/transaction.js:41:38` and see if the problem goes away.

**Comment by aviralgoyal** [comment]
Hi @elhigu 

Thank you for the quick response and it worked!
So I commented out the line you had mentioned and now I'm not getting that unhandled error any more.
And it acquires the connection back after some time. 

Can you please add this bug fix to the library. As I have to install it at several locations and it would be difficult to manually change all the transaction files.

# Question
It takes very long for the connection to stabilize once it has broken around 2 to 3 mins with requests happening every 10 seconds, I assume it will take longer since in production it runs once every 30 seconds. I've added an attachment of the log 
[Log FIle.txt](https://github.com/knex/knex/files/5582564/Log.FIle.txt)
for events below as they happened.

I was searching other links to my problem and I've pasted my knex config and I have commented out the settings suggested by some users. I have't tried them, should I be using any of them or any other recommended settings, will that help stabilize my connection faster?

# Setup
The Application is a syncing application between two setups.
I am running sql server 2017 on two systems:
1. On LAN - Windows 10 - No issues, as soon as I switch off internet it gives connectivity issue and as soon as I switch internet back on it reconnects and starts sending data.
2. On Cloud - Linux Ubuntu 18.04 - I've put a log below. I am connecting to the same sql server instance to two separate databases.

This is the knex config : 
`
const dbData = knex({
    client: "mssql",
    debug: false,
    connection: {
        host: unitIp,
        user: process.env.unitUser,
        password: process.env.unitPass,
        database: process.env.unitDb,
        encrypt: process.env.unitEncrypt === 'false' ? false : true
    },
    pool: {
        min: 1,
        max: 2,
        // createTimeoutMillis: 3000,
        // acquireTimeoutMillis: 30000,
        // idleTimeoutMillis: 30000,
        // reapIntervalMillis: 1000,
        // createRetryIntervalMillis: 100,
       // propagateCreateError: false // <- default is true, set to false
    },
    connectionTimeout: 15000,
});

const unitPool = dbData.client.pool;


const dbWeb = knex({
    client: "mssql",
    debug: false,
    connection: {
        host: process.env.webIp,
        user: process.env.webUser,
        password: process.env.webPass,
        database: process.env.webDb,
        encrypt: process.env.webEncrypt === 'false' ? false : true
    },
    pool: {
        min: 1,
        max: 2,
        // createTimeoutMillis: 3000,
        // acquireTimeoutMillis: 30000,
        // idleTimeoutMillis: 30000,
        // reapIntervalMillis: 1000,
        // createRetryIntervalMillis: 100,
       // propagateCreateError: false // <- default is true, set to false
    },
    connectionTimeout: 15000,
});
const webPool = dbWeb.client.pool;
`

**Comment by aviralgoyal** [comment]
Hi @elhigu 

There is another error I'm getting that is unhandled, and it crashes my application:
Should I comment out the line as well ?

`
node:63876) UnhandledPromiseRejectionWarning: TransactionError: Requests can only be made in the LoggedIn state, not the Final state
    at Request.userCallback (/Users/aviral/projects/texma-sync/node_modules/mssql/lib/tedious/transaction.js:83:24)
    at Request.callback (/Users/aviral/projects/texma-sync/node_modules/tedious/lib/request.js:56:14)
    at Connection.makeRequest (/Users/aviral/projects/texma-sync/node_modules/tedious/lib/connection.js:1800:15)
    at Connection.rollbackTransaction (/Users/aviral/projects/texma-sync/node_modules/tedious/lib/connection.js:1723:17)
    at Immediate._onImmediate (/Users/aviral/projects/texma-sync/node_modules/mssql/lib/tedious/transaction.js:82:32)
    at processImmediate (internal/timers.js:439:21)
<node_internals>/internal/process/warning.js:27
(node:63876) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
<node_internals>/internal/process/warning.js:27
(node:63876) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
`

**Comment by elhigu** [comment]
No. That error is not coming from knex... but good to know too. We should fix these and add some tests to knex to make sure that mssql driver doesn't crash on this kind of unexpected situations.

> Can you please add this bug fix to the library. As I have to install it at several locations and it would be difficult to manually change all the transaction files.

Hopefully someone will submit PR. Right now I don't have time to work on knex bugs.

**Comment by kibertoad** [comment]
@aviralgoyal Could you please send a PR for these issues?

**Comment by kibertoad** [comment]
@elhigu Any ideas how to make rollback throw an error? Can't reproduce the behaviour in test.

**Comment by elhigu** [comment]
Maybe by monkey patching something somewhere or by running a separate query which kills the transaction connection before rollback is ran 🤔  Or by introducing some code to the codebase that only allows testing for that part.

**Comment by aviralgoyal** [comment]
@kibertoad I'll be glad to take the time to send a PR for these issues, but I'm fairly new to all this. If you could guide me as to what all I need to do, I'll might be able to raise it. 

Right now producing this issue was randomly disconnecting my WiFi in the middle of a fairly long running transaction that is reading, deleting and writing to the remote MSSQL db running on an Ubuntu machine maybe 50 read/writes in one transaction. And depending on at what point the transaction was at when disconnecting the WiFi, it would throw a different error and specifically only two of them I found are uncaught as mentioned above. 

So I don't know how to submit a test or PR for this. All I did was comment out that particular line of code in the library and it solved the issue. ( https://github.com/knex/knex/blob/master/lib/dialects/mssql/transaction.js#L41 )

As for the other error https://github.com/knex/knex/issues/4128#issuecomment-732657504 this happens once in 30-40 tries, so again don't know how to reproduce it.

**Comment by kibertoad** [comment]
Fix for original issue released in 0.21.13
